functionary/core/utils/tasking.py

```diff
@@ -83,11 +83,72 @@ def record_task_result(task_result_message: dict) -> None:
         return
 
     TaskLog.objects.create(task=task, log=_protect_output(task, output))
-    TaskResult.objects.create(task=task, result=result)
+    task_result = TaskResult.objects.create(task=task, result=result)
+
+    _update_scheduled_task(task, task_result)
 
     # TODO: This status determination feels like it belongs in the runner. This should
     #       be reworked so that there are explicitly known statuses that could come
     #       back from the runner, rather than passing through the command exit status
     #       as is happening now.
-    task.status = "COMPLETE" if status == 0 else "ERROR"
+    _update_task_status(task, status)
+
+
+@app.task
+def run_scheduled_task(scheduled_task_id: UUID) -> None:
+    """Creates and executes a Task according to a schedule
+
+    Uses the given ScheduledTask object UUID to fetch the ScheduledTask. The necessary
+    metadata is taken from the ScheduledTask object to construct a new Task. The
+    new Task is then associated with that ScheduledTask.
+
+    Args:
+        scheduled_task_id: The UUID of the ScheduledTask object
+
+    Returns:
+        None
+    """
+    scheduled_task = ScheduledTask.objects.get(id=scheduled_task_id)
+
+    _ = Task.objects.create(
+        environment=scheduled_task.environment,
+        creator=scheduled_task.creator,
+        function=scheduled_task.function,
+        parameters=scheduled_task.parameters,
+        scheduled_task=scheduled_task,
+    )
+
+
+def _update_task_status(task: Task, status: int) -> None:
+    match status:
+        case 0:
+            task.status = "COMPLETE"
+        case _:
+            task.status = "ERROR"
+
     task.save()
+
+    if task.scheduled_task is not None and status == "ERROR":
+        task.scheduled_task.error()
+
+
+def _update_scheduled_task(task: Task, task_result: TaskResult) -> None:
+    """Update fields for the scheduled task if available
+
+    If the task is associated with a scheduled task, get the scheduled
+    task reference and update the scheduled task's last run
+    timestamp and most recent task reference.
+
+    Args:
+        task: Task object that the scheduled task will reference as the
+            most recent scheduled task
+        task_result: TaskResult object that is used for task completion timestamp
+
+    Returns:
+        None
+    """
+    if (scheduled_task := task.scheduled_task) is None:
```

Comment:
I originally was using the `PeriodicTask` to get the timestamp of the task that was run. However, they reset that field to `None` whenever the task gets paused, which I disagree'd with.

I usually write conditionals to handle the exit or return case, since I believe it helps reduce the amount of code I would need to nest inside the conditional and looks cleaner. In this case, I want to exit out of that method early if the task is not associated with a `ScheduledTask`, which allows me to leave the rest of the code un-nested that deals with updating the `ScheduledTask`

Reply:
Fair enough.  Carry on.